### PR TITLE
ARROW-11748: [C++] Ensure Decimal fields are in native endian order

### PR DESCRIPTION
**Pull request by cyb70289**

### CODE REVIEW

**Comment by cyb70289** [comment]
No performance difference on little endian machines.

Test result on intel xeon gold 5218.
```
$ archery benchmark diff --suite-filter="arrow-decimal-benchmark" --cc=clang-10 --cxx=clang++-10

----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Non-regressions: (10)
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
              benchmark           baseline          contender  change %                                                                                                               counters
            BinaryBitOp 400.157M items/sec 442.944M items/sec    10.693             {'run_name': 'BinaryBitOp', 'repetitions': 0, 'repetition_index': 0, 'threads': 1, 'iterations': 27844244}
        BinaryMathOp256  56.424M items/sec  56.642M items/sec     0.386          {'run_name': 'BinaryMathOp256', 'repetitions': 0, 'repetition_index': 0, 'threads': 1, 'iterations': 3953034}
        BinaryMathOp128  47.600M items/sec  47.646M items/sec     0.096          {'run_name': 'BinaryMathOp128', 'repetitions': 0, 'repetition_index': 0, 'threads': 1, 'iterations': 3334294}
  BinaryMathOpAggregate  34.017M items/sec  34.044M items/sec     0.080    {'run_name': 'BinaryMathOpAggregate', 'repetitions': 0, 'repetition_index': 0, 'threads': 1, 'iterations': 2381211}
                UnaryOp 404.202M items/sec 404.289M items/sec     0.022                 {'run_name': 'UnaryOp', 'repetitions': 0, 'repetition_index': 0, 'threads': 1, 'iterations': 28295897}
              Constants 265.843M items/sec 265.855M items/sec     0.004               {'run_name': 'Constants', 'repetitions': 0, 'repetition_index': 0, 'threads': 1, 'iterations': 93041786}
        BinaryCompareOp 274.875M items/sec 273.927M items/sec    -0.345         {'run_name': 'BinaryCompareOp', 'repetitions': 0, 'repetition_index': 0, 'threads': 1, 'iterations': 19223572}
               ToString  14.519M items/sec  14.395M items/sec    -0.849                 {'run_name': 'ToString', 'repetitions': 0, 'repetition_index': 0, 'threads': 1, 'iterations': 1694273}
             FromString  17.766M items/sec  17.538M items/sec    -1.282               {'run_name': 'FromString', 'repetitions': 0, 'repetition_index': 0, 'threads': 1, 'iterations': 2074512}
BinaryCompareOpConstant 327.384M items/sec 314.514M items/sec    -3.931 {'run_name': 'BinaryCompareOpConstant', 'repetitions': 0, 'repetition_index': 0, 'threads': 1, 'iterations': 22894103}
```

File: cpp/src/arrow/util/basic_decimal.h [hunk 1]
```diff
@@ -156,8 +162,13 @@ class ARROW_EXPORT BasicDecimal128 {
   static const BasicDecimal128& GetMaxValue();
 
  private:
+#if ARROW_LITTLE_ENDIAN
```
Comment (cyb70289) [inline cpp/src/arrow/util/basic_decimal.h:165]: I think we can use `std::array<uint64_t, 2>` here. Maybe unify `Decimal128` and `Decimal256`.

File: cpp/src/arrow/util/endian.h [hunk 2]
```diff
@@ -169,13 +205,47 @@ static inline T FromBigEndian(T value) {
...
+  const std::array<T, N>& arr;
+
+  explicit LittleEndianArrayReader(const std::array<T, N>& arr) : arr(arr) {}
+
+  const T& operator[](size_t i) const { return arr[ARROW_LITTLE_ENDIAN ? i : N - 1 - i]; }
+};
+
+// Read/write a native endian array as little endian
+template <typename T, size_t N>
+struct LittleEndianArrayWriter {
```
Comment (cyb70289) [inline cpp/src/arrow/util/endian.h:239]: No overhead from these visitor classes, https://godbolt.org/z/3Pn7GMjMo

**Comment by cyb70289** [comment]
@pitrou @bkietz @kiszk

File: cpp/src/arrow/util/endian.h [hunk 3]
```diff
@@ -124,12 +127,23 @@ static inline T ToBigEndian(T value) {
   return ByteSwap(value);
 }
 
+template <typename T, size_t N>
+static inline std::array<T, N> ToBigEndian(std::array<T, N> array) {
```
Comment (pitrou) [inline cpp/src/arrow/util/endian.h:131]: This API is a bit ambiguous. Should it also byteswap the `T`s?

File: cpp/src/arrow/util/endian.h [hunk 4]
```diff
@@ -169,13 +205,47 @@ static inline T FromBigEndian(T value) {
...
+template <typename T, size_t N>
+static inline std::array<T, N> FromLittleEndian(std::array<T, N> array) {
+  std::reverse(array.begin(), array.end());
+  return array;
+}
 #endif
 
+// Read a native endian array as little endian
+template <typename T, size_t N>
+struct LittleEndianArrayReader {
```
Comment (pitrou) [inline cpp/src/arrow/util/endian.h:229]: Perhaps merge this and the writer as a single `LittleEndianArray` class?

File: cpp/src/arrow/util/basic_decimal.h [hunk 5]
```diff
@@ -193,23 +204,25 @@ class ARROW_EXPORT BasicDecimal256 {
   static constexpr int bit_width = 256;
 
   /// \brief Create a BasicDecimal256 from the two's complement representation.
-  constexpr BasicDecimal256(const std::array<uint64_t, 4>& little_endian_array) noexcept
-      : little_endian_array_(little_endian_array) {}
+  constexpr BasicDecimal256(const std::array<uint64_t, 4>& array) noexcept
+      : array_(array) {}
```
Comment (pitrou) [inline cpp/src/arrow/util/basic_decimal.h:208]: Add a docstring comment about endianness?

File: cpp/src/arrow/util/basic_decimal.cc [hunk 6]
```diff
@@ -572,20 +593,23 @@ struct uint128_t {
...
 // Multiplies two N * 64 bit unsigned integer types, represented by a uint64_t
 // array into a same sized output. Elements in the array should be in
-// little endian order, and output will be the same. Overflow in multiplication
+// native endian order, and output will be the same. Overflow in multiplication
 // will result in the lower N * 64 bits of the result being set.
 template <int N>
 inline void MultiplyUnsignedArray(const std::array<uint64_t, N>& lh,
                                   const std::array<uint64_t, N>& rh,
                                   std::array<uint64_t, N>* result) {
+  BitUtil::LittleEndianArrayReader<uint64_t, N> lh_le(lh), rh_le(rh);
```
Comment (pitrou) [inline cpp/src/arrow/util/basic_decimal.cc:602]: `const`?

File: cpp/src/arrow/util/basic_decimal.cc [hunk 7]
```diff
@@ -121,217 +121,238 @@ static const BasicDecimal128 ScaleMultipliersHalf[] = {
     BasicDecimal128(271050543121376108LL, 9257742014424809472ULL),
     BasicDecimal128(2710505431213761085LL, 343699775700336640ULL)};
 
+#define BasicDecimal256FromLE(v1, v2, v3, v4) \
+  BasicDecimal256(BitUtil::FromLittleEndian<uint64_t, 4>(v1, v2, v3, v4))
```
Comment (pitrou) [inline cpp/src/arrow/util/basic_decimal.cc:125]: Can you `undef` this after?

**Review by pitrou — COMMENTED** [review]
+1 on the principle. A couple small comments.

File: cpp/src/arrow/util/basic_decimal.cc (hunk 7 above)
Comment (cyb70289) [reply cpp/src/arrow/util/basic_decimal.cc]: Done

File: cpp/src/arrow/util/basic_decimal.cc (hunk 6 above)
Comment (cyb70289) [reply cpp/src/arrow/util/basic_decimal.cc]: Done

File: cpp/src/arrow/util/basic_decimal.h (hunk 5 above)
Comment (cyb70289) [reply cpp/src/arrow/util/basic_decimal.h:209]: Added

File: cpp/src/arrow/util/endian.h (hunk 3 above)
Comment (cyb70289) [reply cpp/src/arrow/util/endian.h]: Removed this API with general name.

File: cpp/src/arrow/util/endian.h (hunk 4 above)
Comment (cyb70289) [reply cpp/src/arrow/util/endian.h]: Looks not easy to combine the Reader and Writer classes. Writer can do both reading and writing, it accepts a pointer. Reader accepts const reference. The `const T& operator []` is hard to merge.

I added a LittleEndianArray namespace and moved everything in. Export only `Make` function to construct Reader or Writer per input type.

**Review by pitrou — APPROVED** [review]
+1, thank you @cyb70289

**Pull request closed without merging**